docs/isl-2-0/spec.md

```diff
@@ -0,0 +1,674 @@
+
+# Types
+
+The Ion Schema type system is a hybrid of [nominal](https://en.wikipedia.org/wiki/Nominal_type_system) and [structural](https://en.wikipedia.org/wiki/Structural_type_system) typing.
+All Ion values are nominally one of the types defined in the [Ion data model](https://amzn.github.io/ion-docs/docs/spec.html#the-ion-data-model).
+Beyond that, a value (or stream of values) may belong to one or more structural types.
+(All values belong to at least one structural type—the top type `$any`.)
+
+## Built-in Types
+
+Ion Schema has several built-in types that are implicitly imported before any other imports are handled.
+
+The nominal Ion types are prefixed with `$`, and correspond precisely with the types defined by the Ion data model, including strongly-typed null values:
+- **scalars:** `$blob`, `$bool`, `$clob`, `$decimal`, `$float`, `$int`, `$null`,
+  `$string`, `$symbol`, `$timestamp`
+- **containers:** `$list`, `$sexp`, `$struct`
+
+Ion Schema adds one more nominal type, `document`, which is a stream of top-level Ion values.
+Ion Schema implementations may represent this stream as a list, sequence, array, or other similar data structure depending on what is idiomatic for the language in which it is implemented.
+
+In addition, Ion Schema provides the following structural types:
+
+- `$any`: the [top type](https://en.wikipedia.org/wiki/Top_type); every value is valid
+- `$lob`: represents a `$blob` or `$clob`
+- `$number`: represents a `$decimal`, `$float`, or `$int`
+- `$text`: represents a `$string` or `$symbol`
+
+- `any`: represents any value that is not a null value
+- `blob`, `bool`, `clob`, `decimal`, `float`, `int`, `string`, `symbol`, `timestamp`, `list`, `sexp`, `struct` correspond to the Ion types, except that they do not allow null values.
+- `lob`: represents a (non-null) `blob` or `clob`
+- `number`: represents a (non-null) `decimal`, `float`, or `int`
+- `text`: represents a (non-null) `string` or `symbol`
+- `nothing`: the [empty type](https://en.wikipedia.org/wiki/Empty_type); it has no valid value
+
+{% include note.html type="tip" content="
+For the built-in types, the presence of a leading `$` sigil indicates that the type includes null values.
+" %}
+
+## Type Definitions
+
+{% include grammar-element.md productions="named_type_definition,inline_type_definition,variably_occurring_type_reference" %}
+
+A type consists of a collection of zero or more constraints.
+The set of values which belong to a type is the intersection of the values that satisfy each constraint.
+In order for a value to be a valid instance of a type, the value must not violate any of the type's constraints.
+A type definition with no constraints is equivalent to `$any`.
+
+{% include example.md title="The definition of `$any`, the top type" markdown="
+```ion
+type::{
+  name: $any,
+}
+```
+" %}
+
+Top-level types must declare a `name`. All other type definitions must not declare a `name`.
```

Comment:
(Suggestion) I see that you have already declared top type in the above section for built-in types. But this does create slight confusion between top type and top-level type. 
I think we should either use the word top type or top-level type everywhere.

Reply:
Thanks for pointing this out. A "top-level type" is actually a different thing than "the top type". I'll see about adding some explanatory notes or changing some terms so that it's clear.

Reply:
I've updated the spec to use "universal type" instead of "top type". (They are synonyms, but I think "universal type" is a better choice to avoid confusion here.)